SparseGrids/tsgRuleLocalPolynomial.hpp

```diff
@@ -261,6 +265,86 @@ class templRuleLocalPolynomial : public BaseRuleLocalPolynomial{
             }
         }
     }
+
+    double diffRaw(int point, double x) const override{
+        if (isZeroOrder) {
+            return 0.0;
+        } else {
+            if (rule == rule_localp or rule == rule_semilocalp) {
+                if (point == 0) return 0.0;
+                if (rule == rule_semilocalp) {
+                    if (point == 1) return x - 0.5;
+                    if (point == 2) return x + 0.5;
+                }
+            }
+            double xn = scaleX(point, x);
+            double an = scaleDiffX(point);
+            if (rule != rule_semilocalp and max_order == 1) {
+                if (x == 1.0) {
+                    if (rule == rule_localp0 and point == 0)
+                        return  -1.0;
+                    else if (rule == rule_localpb and point == 1)
+                        return 0.5;
+                    else if (rule == rule_localpb and point == 2)
+                        return -1.0;
+                    else if (point == 2)
+                        return an;
+                }
+                return (xn >= 0 ? -1.0 : 1.0) * an;
+            }
+            if (max_order == 2) return an * diffPWQuadratic(point, xn);
+            if (max_order == 3) return an * diffPWCubic(point, xn);
+            return an * diffPWPower(point, xn);
+        }
+    }
+
+    double diffSupport(int point, double x, bool &isSupported) const override{
+        if (isZeroOrder) {
+            isSupported = false;
+            return 0.0;
+        } else {
+            // We still need (isSupported == true) at point = 0 so that TasGrid::GridLocalPolynomial::walkTree<3>()
+            // walks through the children of this point.
+           isSupported = true;
+            if (rule == rule_localp or rule == rule_semilocalp) {
+                if (point == 0) {
+                    return 0.0;
+                }
+                if (rule == rule_semilocalp) {
+                    if (point == 1) return x - 0.5;
+                    if (point == 2) return x + 0.5;
+                }
+            }
+            // Note that the support will be of the form [a, b), except for x = +1.0, i.e., the rightmost node in the domain. This
+            // is to avoid double-counting derivatives at points of discontinuity.
+            double xn = scaleX(point, x);
+            bool rightmost_node = (x == 1.0 and xn == 1.0);
+            isSupported = (-1.0 <= xn and xn < 1.0) or rightmost_node;
+            if (isSupported) {
+                double an = scaleDiffX(point);
+                if (rule != rule_semilocalp and max_order == 1) {
+                    // Edge cases due to the logic of avoiding double-counting.
+                    if (x == 1.0) {
```

Comment:
The first three if-statements are not needed, you would get the exact same result if you just apply the return formula on line `337`. I'm not sure about the 4th statement, but I wonder why that would be needed.

Reply:
You're right. The first three are redundant, so I have removed them. 

We can see that the fourth one is necessary for `TasGrid::rule_localp`, `x == 1.0`, and `point == 2` by tracing the following facts: 

- `xn == 0.0` from the logic in `scaleX()`
- `an == 1.0` from the logic in `scaleDiffX()` and the value of `xn`
- `(xn >= 0 ? -1.0 : 1.0) * an == -1.0` from the above two statements
- at the boundary where `x == 1.0` only the left derivative exists 
- since the left derivative at `point == 2` for `xn == 0.0` is `1.0`, we should not take the logic in line `337` and hardcode it as `an`